test/e2e/testsuites/dynamically_provisioned_garbage_collection_replicas_tester.go

```diff
@@ -0,0 +1,164 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testsuites
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
+	diskv1alpha2 "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/azuredisk/v1alpha2"
+	azDiskClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	"sigs.k8s.io/azuredisk-csi-driver/test/e2e/driver"
+	testtypes "sigs.k8s.io/azuredisk-csi-driver/test/types"
+)
+
+// DynamicallyProvisionedScaleReplicasOnDetach will provision required PV(s), PVC(s) and Pod(s), enough to fill the available
+// spots for attachments on all the nodes. Another PVC will be added that will have to wait for a opening on a node to connect
+
+type DynamicallyProvisionedScaleReplicasOnDetach struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	StatefulSetPod         testtypes.PodDetails
+	Replicas               int
+	StorageClassParameters map[string]string
+	AzDiskClient           *azDiskClientSet.Clientset
+	NewPod                 testtypes.PodDetails
+}
+
+func (t *DynamicallyProvisionedScaleReplicasOnDetach) Run(client clientset.Interface, namespace *v1.Namespace, schedulerName string) {
+
+	//Cordon all schedulable nodes except for the two we are running our test on: one for primary attachments, one for replicas
+	nodes, err := client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{FieldSelector: fields.Set{
+		"spec.unschedulable": "false",
+	}.AsSelector().String()})
+	framework.ExpectNoError(err)
+
+	if len(nodes.Items) < 3 {
+		ginkgo.Skip("need at least 3 schedulable nodes to verify the test case. Current node count is %d", len(nodes.Items))
+	}
+
+	testPod := testtypes.TestPod{
+		Client: client,
+	}
+	for i, node := range nodes.Items {
+		if i > 1 {
+			testPod.SetNodeUnschedulable(node.Name, true)
+		}
+	}
+
+	var tStatefulSet *testtypes.TestStatefulset
+	var wg sync.WaitGroup
+	var tokens = make(chan struct{}, 20) // avoid too many concurrent requests
+	tStorageClass, storageCleanup := t.StatefulSetPod.CreateStorageClass(client, namespace, t.CSIDriver, t.StorageClassParameters)
+	defer storageCleanup()
+	tStatefulSet, cleanupStatefulSet := t.StatefulSetPod.SetupStatefulset(client, namespace, t.CSIDriver, schedulerName, t.Replicas, t.StorageClassParameters, &tStorageClass)
+
+	wg.Add(1)
+	go func(ss *testtypes.TestStatefulset) {
```

Comment:
Why dispatch the creation of the single StatefulSet to a goroutine? It seems like overkill.